File: SuperiorPackGroup/BLL/ShippingReturnDetailsBLL.vb
Imports SuperiorPackGroup.SPGTableAdapters
Imports System.Text
Imports DevExpress.Xpo
Imports DXDAL.SPGData
Imports DevExpress.Data.Filtering

<ComponentModel.DataObject()> _
Public Class ShippingReturnDetailsBLL

    Private m_ShippingReturnDetailsTableAdapter As ShippingReturnDetailsTableAdapter = Nothing
    Public ReadOnly Property Adapter() As ShippingReturnDetailsTableAdapter

        Get
            If m_ShippingReturnDetailsTableAdapter Is Nothing Then
                m_ShippingReturnDetailsTableAdapter = New ShippingReturnDetailsTableAdapter()
                m_ShippingReturnDetailsTableAdapter.Connection.ConnectionString = My.Settings.UserConnectionString
            End If
            Return m_ShippingReturnDetailsTableAdapter
        End Get

    End Property

    Private Sub UpdateAuditTrail(ByVal ModifiedRecord As SPG.ShippingReturnDetailsRow, ByVal originalRecord As Object())

        Dim builder As New StringBuilder(String.Empty)
        Dim recordFields As Integer = (ModifiedRecord.ItemArray.Length - 1)
        Dim i As Integer = 0
        Do While (i <= recordFields)
            Try
                If IsDBNull(originalRecord(i)) Then
                    If Not IsDBNull(ModifiedRecord.Item(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, "NULL", ModifiedRecord.Item(i)))
                    End If
                ElseIf IsDBNull(ModifiedRecord.Item(i)) Then
                    If Not IsDBNull(originalRecord(i)) Then
                        builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), "NULL"))
                    End If
                ElseIf ModifiedRecord.Item(i) IsNot originalRecord(i) Then
                    builder.Append(String.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns.Item(i).ColumnName, originalRecord(i), ModifiedRecord.Item(i)))
                End If
            Catch
            End Try
            i += 1
        Loop
        If (builder.Length > 2) Then
            builder.Length = (builder.Length - 2)
            AuditTrailBLL.AddTrailEntry(My.Settings.UserName, CType(ModifiedRecord.Item(0), Integer), "ShippingReturnDetails", builder.ToString)
        End If

    End Sub

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Select, False)> _
    Public Function GetShippingReturnDetailsByItemID(ByVal itemID As Integer) As SPG.ShippingReturnDetailsDataTable

        Return Adapter.GetShippingReturnDetailsByItemID(itemID)

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Select, False)> _
    Public Function GetShippingReturnDetailsByShippingID(ByVal shippingID As Integer) As SPG.ShippingReturnDetailsDataTable

        Return Adapter.GetShippingReturnDetailsByShippingID(shippingID)

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Update, True)> _
    Public Function UpdateShippingReturnDetails(ByVal session As Session, ByVal detailID As Integer?, ByVal receivingID As Integer, ByVal itemID As Integer?, ByVal lot As String, ByVal quantity As Integer?, ByVal units As Integer?,
                                            ByVal pallets As Single?, ByVal reason As String, ByVal expirationDate As Date?) As Boolean

        If Not itemID.HasValue Then
            Throw New ApplicationException("You must provide return item.")
        End If

        If Not quantity.HasValue OrElse Not units.HasValue Then
            Throw New ApplicationException("You must provide the Quantity\Units returned.")
        End If

        Dim details As SPG.ShippingReturnDetailsDataTable = Adapter.GetShippingReturnDetailByID(detailID.Value)

        If details.Count = 0 Then
            'It is a new Detail
            Return Me.InsertDetails(session, receivingID, itemID, lot, quantity, units, pallets, reason, expirationDate)
        End If

        Dim returnDetail As SPG.ShippingReturnDetailsRow = details(0)
        Dim originalItem, originalQuantity, newQuantity As Integer
        Dim itemChanged As Boolean = False

        Dim originalRecord As Object() = Nothing
        originalRecord = returnDetail.ItemArray

        originalItem = returnDetail.ReturnDetItemID
        If originalItem <> itemID.Value Then
            itemChanged = True
        End If
        returnDetail.ReturnDetItemID = itemID.Value
        If String.IsNullOrEmpty(lot) Then
            returnDetail.SetReturnDetLotNull()
        Else
            returnDetail.ReturnDetLot = lot
        End If
        returnDetail.ReturnDetQty = quantity.Value
        originalQuantity = returnDetail.intUnits
        newQuantity = returnDetail.intUnits - units.Value
        returnDetail.intUnits = units.Value
        If Not pallets.HasValue Then
            returnDetail.SetsngPalletsNull()
        Else
            returnDetail.sngPallets = pallets.Value
        End If
        If String.IsNullOrEmpty(reason) Then
            returnDetail.SetReasonNull()
        Else
            returnDetail.Reason = reason
        End If
        If expirationDate.HasValue Then
            returnDetail.ExpirationDate = expirationDate.Value
        Else
            returnDetail.SetExpirationDateNull()
        End If

        If Not IsNothing(originalRecord) Then
            Me.UpdateAuditTrail(returnDetail, originalRecord)
        End If

        Dim rowsAffected As Integer = Adapter.Update(returnDetail)

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            Dim locationID As Integer = session.GetObjectByKey(Of Shipping)(receivingID).ShippingLocation.Oid
            If itemChanged = True Then
                items.UpdateStock(session, originalItem, originalQuantity, False, locationID)
                items.UpdateStock(session, itemID.Value, units.Value * -1, False, locationID)
            Else
                If newQuantity <> 0 Then
                    items.UpdateStock(session, itemID.Value, newQuantity, False, locationID)
                End If
            End If
        End If

        Return rowsAffected = 1

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Insert, True)> _
    Public Function InsertDetails(ByVal session As Session, ByVal shippingID As Integer, ByVal itemID As Integer?, ByVal lot As String, ByVal quantity As Integer?, ByVal units As Integer?, ByVal pallets As Single?, _
                                  ByVal reason As String, ByVal expirationDate As Date?) As Boolean

        Dim details As SPG.ShippingReturnDetailsDataTable = New SPG.ShippingReturnDetailsDataTable
        Dim returnDetail As SPG.ShippingReturnDetailsRow = details.NewShippingReturnDetailsRow()

        returnDetail.ShipMainID = shippingID
        returnDetail.ReturnDetItemID = itemID.Value
        If String.IsNullOrEmpty(lot) Then
            returnDetail.SetReturnDetLotNull()
        Else
            returnDetail.ReturnDetLot = lot
        End If
        returnDetail.ReturnDetQty = quantity.Value
        returnDetail.intUnits = units.Value
        If Not pallets.HasValue Then
            returnDetail.SetsngPalletsNull()
        Else
            returnDetail.sngPallets = pallets.Value
        End If
        If String.IsNullOrEmpty(reason) Then
            returnDetail.SetReasonNull()
        Else
            returnDetail.Reason = reason
        End If
        If expirationDate.HasValue Then
            returnDetail.ExpirationDate = expirationDate.Value
        Else
            returnDetail.SetExpirationDateNull()
        End If

        details.AddShippingReturnDetailsRow(returnDetail)
        Dim rowsAffected As Integer = Adapter.Update(details)

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            items.UpdateStock(session, itemID.Value, units.Value * -1, False, session.GetObjectByKey(Of Shipping)(shippingID).ShippingLocation.Oid)
        End If

        Return rowsAffected = 1

    End Function

    <ComponentModel.DataObjectMethod(ComponentModel.DataObjectMethodType.Delete, True)> _
    Public Function DeleteReturnDetail(ByVal session As Session, ByVal detailID As Integer) As Boolean

        Dim details As SPG.ShippingReturnDetailsDataTable = Adapter.GetShippingReturnDetailByID(detailID)
        Dim rowsAffected As Integer = 0
        Dim itemID, quantity, locationID As Integer

        If details.Count = 1 Then
            Dim returnDetail As SPG.ShippingReturnDetailsRow = CType(details.Rows(0), SPG.ShippingReturnDetailsRow)
            itemID = returnDetail.ReturnDetItemID
            quantity = returnDetail.intUnits
            locationID = session.GetObjectByKey(Of Shipping)(returnDetail.ShipMainID).ShippingLocation.Oid
            rowsAffected = Adapter.Delete(detailID, returnDetail.ts)
        End If

        If rowsAffected = 1 Then
            Dim items As ItemsBLL = New ItemsBLL
            items.UpdateStock(session, itemID, quantity, False, locationID)
        End If

        'Return true if precisely one row was deleted, otherwise return false.
        Return rowsAffected = 1

    End Function

    Public Shared Function GetShippingReturnDetailsViewByItemID(ByVal itemID As Integer) As XPView

        Dim ShippingReturnXPView As New XPView(Session.DefaultSession, GetType(ShippedReturns))

        ShippingReturnXPView.Properties.AddRange(New ViewProperty() {New ViewProperty("ReturnDetID", SortDirection.None, ShippedReturns.Fields.ReturnDetID, False, True),
                                                                     New ViewProperty("ShipMainID", SortDirection.Ascending, ShippedReturns.Fields.ShipMainID.ShipMainID, False, True),
                                                                     New ViewProperty("ReturnDetLot", SortDirection.None, ShippedReturns.Fields.ReturnDetLot, False, True),
                                                                     New ViewProperty("ExpirationDate", SortDirection.None, ShippedReturns.Fields.ExpirationDate, False, True),
                                                                     New ViewProperty("intUnits", SortDirection.None, ShippedReturns.Fields.intUnits, False, True),
                                                                     New ViewProperty("ShipMainBOL", SortDirection.None, ShippedReturns.Fields.ShipMainID.ShipMainBOL, False, True),
                                                                     New ViewProperty("ShipMainDate", SortDirection.None, ShippedReturns.Fields.ShipMainID.ShipMainDate, False, True)})

        ShippingReturnXPView.Criteria = New BinaryOperator(ShippedReturns.Fields.ReturnDetItemID.ItemID.PropertyName, itemID, BinaryOperatorType.Equal)

        Return ShippingReturnXPView

    End Function

End Class

File: SuperiorPackGroup/BLL/ShippingReturnDetailsBLL.cs
//INSTANT C# NOTE: Formerly VB project-level imports:
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Drawing;
using System.Diagnostics;
using System.Windows.Forms;

using SuperiorPackGroup.SPGTableAdapters;
using System.Text;
using DevExpress.Xpo;
using DXDAL.SPGData;
using DevExpress.Data.Filtering;
using System.Linq;

namespace SuperiorPackGroup
{
	[System.ComponentModel.DataObject()]
	public class ShippingReturnDetailsBLL
	{

		private ShippingReturnDetailsTableAdapter m_ShippingReturnDetailsTableAdapter = null;
		public ShippingReturnDetailsTableAdapter Adapter
		{

			get
			{
				if (m_ShippingReturnDetailsTableAdapter == null)
				{
					m_ShippingReturnDetailsTableAdapter = new ShippingReturnDetailsTableAdapter();
					m_ShippingReturnDetailsTableAdapter.Connection.ConnectionString = Properties.Settings.Default.UserConnectionString;
				}
				return m_ShippingReturnDetailsTableAdapter;
			}

		}

		private void UpdateAuditTrail(SPG.ShippingReturnDetailsRow ModifiedRecord, object[] originalRecord)
		{

			StringBuilder builder = new StringBuilder(string.Empty);
			int recordFields = (ModifiedRecord.ItemArray.Length - 1);
			int i = 0;
			while (i <= recordFields)
			{
				try
				{
					if (Convert.IsDBNull(originalRecord[i]))
					{
						if (!Convert.IsDBNull(ModifiedRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, "NULL", ModifiedRecord[i]));
						}
					}
					else if (Convert.IsDBNull(ModifiedRecord[i]))
					{
						if (!Convert.IsDBNull(originalRecord[i]))
						{
							builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], "NULL"));
						}
					}
					else if (ModifiedRecord[i] != originalRecord[i])
					{
						builder.Append(string.Format("{0}:{1}({2}); ", ModifiedRecord.Table.Columns[i].ColumnName, originalRecord[i], ModifiedRecord[i]));
					}
				}
				catch
				{
				}
				i += 1;
			}
			if (builder.Length > 2)
			{
				builder.Length = (builder.Length - 2);
				AuditTrailBLL.AddTrailEntry(Properties.Settings.Default.UserName, Convert.ToInt32(ModifiedRecord[0]), "ShippingReturnDetails", builder.ToString());
			}

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, false)]
		public SPG.ShippingReturnDetailsDataTable GetShippingReturnDetailsByItemID(int itemID)
		{

			return Adapter.GetShippingReturnDetailsByItemID(itemID);

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Select, false)]
		public SPG.ShippingReturnDetailsDataTable GetShippingReturnDetailsByShippingID(int shippingID)
		{

			return Adapter.GetShippingReturnDetailsByShippingID(shippingID);

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Update, true)]
		public bool UpdateShippingReturnDetails(Session session, int? detailID, int receivingID, int? itemID, string lot, int? quantity, int? units, float? pallets, string reason, DateTime? expirationDate)
		{

			if (!itemID.HasValue)
			{
				throw new ApplicationException("You must provide return item.");
			}

			if (!quantity.HasValue || !units.HasValue)
			{
				throw new ApplicationException("You must provide the Quantity\\Units returned.");
			}

			SPG.ShippingReturnDetailsDataTable details = Adapter.GetShippingReturnDetailByID(detailID.Value);

			if (details.Count() == 0)
			{
				//It is a new Detail
				return InsertDetails(session, receivingID, itemID, lot, quantity, units, pallets, reason, expirationDate);
			}

			SPG.ShippingReturnDetailsRow returnDetail = details[0];
			int originalItem = 0;
			int originalQuantity = 0;
			int newQuantity = 0;
			bool itemChanged = false;

			object[] originalRecord = returnDetail.ItemArray;

			originalItem = returnDetail.ReturnDetItemID;
			if (itemID.HasValue && originalItem != itemID.Value)
			{
				itemChanged = true;
			}
			returnDetail.ReturnDetItemID = itemID.Value;
			if (string.IsNullOrEmpty(lot))
			{
				returnDetail.SetReturnDetLotNull();
			}
			else
			{
				returnDetail.ReturnDetLot = lot;
			}
			returnDetail.ReturnDetQty = quantity.Value;
			originalQuantity = returnDetail.intUnits;
			newQuantity = returnDetail.intUnits - units.Value;
			returnDetail.intUnits = units.Value;
			if (!pallets.HasValue)
			{
				returnDetail.SetsngPalletsNull();
			}
			else
			{
				returnDetail.sngPallets = pallets.Value;
			}
			if (string.IsNullOrEmpty(reason))
			{
				returnDetail.SetReasonNull();
			}
			else
			{
				returnDetail.Reason = reason;
			}
			if (expirationDate.HasValue)
			{
				returnDetail.ExpirationDate = expirationDate.Value;
			}
			else
			{
				returnDetail.SetExpirationDateNull();
			}

			if (!(originalRecord == null))
			{
				UpdateAuditTrail(returnDetail, originalRecord);
			}

			int rowsAffected = Adapter.Update(returnDetail);

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				int locationID = session.GetObjectByKey<Shipping>(receivingID).ShippingLocation.Oid;
				if (itemChanged == true)
				{
					items.UpdateStock(session, originalItem, originalQuantity, false, locationID);
					items.UpdateStock(session, itemID.Value, units.Value * -1, false, locationID);
				}
				else
				{
					if (newQuantity != 0)
					{
						items.UpdateStock(session, itemID.Value, newQuantity, false, locationID);
					}
				}
			}

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Insert, true)]
		public bool InsertDetails(Session session, int shippingID, int? itemID, string lot, int? quantity, int? units, float? pallets, string reason, DateTime? expirationDate)
		{

			SPG.ShippingReturnDetailsDataTable details = new SPG.ShippingReturnDetailsDataTable();
			SPG.ShippingReturnDetailsRow returnDetail = details.NewShippingReturnDetailsRow();

			returnDetail.ShipMainID = shippingID;
			returnDetail.ReturnDetItemID = itemID.Value;
			if (string.IsNullOrEmpty(lot))
			{
				returnDetail.SetReturnDetLotNull();
			}
			else
			{
				returnDetail.ReturnDetLot = lot;
			}
			returnDetail.ReturnDetQty = quantity.Value;
			returnDetail.intUnits = units.Value;
			if (!pallets.HasValue)
			{
				returnDetail.SetsngPalletsNull();
			}
			else
			{
				returnDetail.sngPallets = pallets.Value;
			}
			if (string.IsNullOrEmpty(reason))
			{
				returnDetail.SetReasonNull();
			}
			else
			{
				returnDetail.Reason = reason;
			}
			if (expirationDate.HasValue)
			{
				returnDetail.ExpirationDate = expirationDate.Value;
			}
			else
			{
				returnDetail.SetExpirationDateNull();
			}

			details.AddShippingReturnDetailsRow(returnDetail);
			int rowsAffected = Adapter.Update(details);

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				items.UpdateStock(session, itemID.Value, units.Value * -1, false, session.GetObjectByKey<Shipping>(shippingID).ShippingLocation.Oid);
			}

			return rowsAffected == 1;

		}

		[System.ComponentModel.DataObjectMethod(System.ComponentModel.DataObjectMethodType.Delete, true)]
		public bool DeleteReturnDetail(Session session, int detailID)
		{

			SPG.ShippingReturnDetailsDataTable details = Adapter.GetShippingReturnDetailByID(detailID);
			int rowsAffected = 0;
			int itemID = 0;
			int quantity = 0;
			int locationID = 0;

			if (details.Count() == 1)
			{
				SPG.ShippingReturnDetailsRow returnDetail = (SPG.ShippingReturnDetailsRow)details.Rows[0];
				itemID = returnDetail.ReturnDetItemID;
				quantity = returnDetail.intUnits;
				locationID = session.GetObjectByKey<Shipping>(returnDetail.ShipMainID).ShippingLocation.Oid;
				rowsAffected = Adapter.Delete(detailID, returnDetail.ts);
			}

			if (rowsAffected == 1)
			{
				ItemsBLL items = new ItemsBLL();
				items.UpdateStock(session, itemID, quantity, false, locationID);
			}

			//Return true if precisely one row was deleted, otherwise return false.
			return rowsAffected == 1;

		}

		public static XPView GetShippingReturnDetailsViewByItemID(int itemID)
		{

			XPView ShippingReturnXPView = new XPView(Session.DefaultSession, typeof(ShipDet));

			ShippingReturnXPView.Properties.AddRange(new ViewProperty[]
			{
				new ViewProperty("ReturnDetID", SortDirection.None, ShipDet.Fields.ShipDetDetID, false, true),
				new ViewProperty("ShipMainID", SortDirection.Ascending, ShipDet.Fields.ShipDetMainID.ShipMainID, false, true),
				new ViewProperty("ReturnDetLot", SortDirection.None, ShipDet.Fields.ShipDetLot, false, true),
				new ViewProperty("ExpirationDate", SortDirection.None, ShipDet.Fields.ExpirationDate, false, true),
				new ViewProperty("intUnits", SortDirection.None, ShipDet.Fields.ShipDetDetQty, false, true),
				new ViewProperty("ShipMainBOL", SortDirection.None, ShipDet.Fields.ShipDetMainID.ShipMainBOL, false, true),
				new ViewProperty("ShipMainDate", SortDirection.None, ShipDet.Fields.ShipDetMainID.ShipMainDate, false, true)
			});

			ShippingReturnXPView.Criteria = CriteriaOperator.And(new BinaryOperator(ShipDet.Fields.ShipDetItemID.ItemType.PropertyName, "FG", BinaryOperatorType.NotEqual), new BinaryOperator(ShipDet.Fields.ShipDetItemID.ItemID.PropertyName, itemID, BinaryOperatorType.Equal));

			return ShippingReturnXPView;

		}

	}

}